### ISSUE TITLE
LMDB RAM consumption is not restricted

### ISSUE THREAD

**Issue by ManyTheFish** [meilisearch/meilisearch#4764]
Related to https://github.com/meilisearch/meilisearch/issues/3397

## Summary

When indexing or changing major settings on a big dataset on a Meilisearch instance, restricting the indexing memory using the flag `--max-indexing-memory 2GB`, the RAM consumption is most of the time higher than the threshold:

    8GB wiki:
        full indexing: spike at 5.37GB
        change filter setting: spike at 3.02GB
        change the searchable setting: pike at 2.47GB but 7.85GB when writing words_prefix_integer_docids
    1GB songs:
        full indexing: spike at 5.00GB
        change filter setting: spike at 1.91GB
        change the searchable setting: pike at 3.44GB but 4.32GB when writing words_prefix_integer_docids

In the [previous issue](https://github.com/meilisearch/meilisearch/issues/3397), this overuse was due to the extractors. It doesn't seem to be the case anymore, but it is due to LMDB writings. 
LMDB has no restrictions in terms of memory usage for now. However, it's not because LMDB uses a lot of RAM on a big machine (more than 8GB of RAM) that it would make a smaller machine crash.

> Next steps:
> - test the same use case on a machine with 4GB of maximum RAM and see if the indexing fails
> - investigate a way to constrain the LMDB RAM consumption

**Comment by camgaertner** [comment]
I think I'm affected by this as well. Is there any potential workaround for the time being? I upgraded from 1.7 to 1.9 hoping it would help, but still encountering the issue.

Tried setting max-indexing memory to `500 Mb`, but it exceeds to 4GB+ and ends up getting OOM-killed.
```
[88696.470095] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=meilisearch.service,mems_allowed=0,global_oom,task_memcg=/system.slice/meilisearch.service,task=meilisearch,pid=17149,uid=993
[88696.470156] Out of memory: Killed process 17149 (meilisearch) total-vm:4321940096kB, anon-rss:2289336kB, file-rss:2668kB, shmem-rss:0kB, UID:993 pgtables:37864kB oom_score_adj:0
```

For now, I'm experimenting with providing smaller batches for document updates as well as waiting for every task to finish before submitting a new one.

**Comment by dureuill** [comment]
Hello @camgaertner :wave:,

1. can you tell us more about your environment? OS, Meilisearch version
2. can you retry with `--experimental-reduce-indexing-memory-usage`?

Apologies for the late answer

**Comment by camgaertner** [comment]
> Hello @camgaertner 👋,
> 
>     1. can you tell us more about your environment? OS, Meilisearch version
> 
>     2. can you retry with `--experimental-reduce-indexing-memory-usage`?
> 
>     3. are you using the vector store?
> 
>     4. how many documents do you have?
> 
> 
> Apologies for the late answer

Hi @dureuill , thanks for following up. I think my issue was caused partially by another program running and using memory while Meilisearch was indexing, but it seems like `max-indexing memory` was not helping to lower the memory usage in Meilisearch either.

I addressed the issue by reducing memory usage in the other program, and Meilisearch mostly stopped getting OOM-killed. I'll try `--experimental-reduce-indexing-memory-usage` if I have this problem again in the future.

If you are still interested in this info:
OS: Ubuntu 24.04
Meilisearch version: 1.9.0
Docs: ~370k

**Comment by dureuill** [comment]
Thank you for your answer, I'm glad you found a fix to your issue

**Comment by YoEight** [comment]
I conducted an investigation on this.

**TL;DR:** On a 4GB machine, using `--max-indexing-memory 2GB` will cause RAM consumption to exceed 2GB, but it will not cause indexing to fail.

---

`max-indexing-memory` only covers the memory used by `meilisearch`. LMDB memory consumption is not capped, and currently there is no setting to do so.

Using `MDB_WRITEMAP` has no impact in this case. LMDB is fully memory-mapped, and its behavior is largely controlled by the OS.

The core issue is LMDB causing a large number of dirty pages in resident memory, despite LMDB having committed the data. Those pages are "logically" clean from LMDB's point of view since they were persisted to disk successfully.

Those pages remain dirty in the OS because the kernel delays write-back aggressively when memory pressure is low. I tested this claim by creating a large index on two different machines:

- One with 96GB of RAM
- One with 4GB of RAM

On the beefy machine, the OS doesn't reclaim those dirty pages at all because memory is plentiful. On the less capable machine, the OS reclaims those dirty pages aggressively.

My current assumption is that whatever LMDB is doing when committing data doesn't clear the dirty flag on the mmap'd pages. Those virtual pages stay dirty because they contain modified data that differs from the original disk contents even thought that modified data has been persisted.
